source/MaterialXView/ShaderValidators/Glsl/GlslValidator.h

```diff
@@ -0,0 +1,326 @@
+///
+class GlslValidator
+{
+  public:
+    /// Structure to hold information about program inputs
+    struct ProgramInput
+    {
+        static int INVALID_OPENGL_TYPE;
+
+        /// Program location. -1 means an invalid location
+        int _location;
+        /// OpenGL type of the input. -1 means an invalid type
+        int _type;
+        /// Size.
+        int _size;
+
+        /// Program input constructor
+        ProgramInput(int inputLocation, int inputType, int inputSize)
+        : _location(inputLocation)
+        , _type(inputType)
+        , _size(inputSize)
+        {}
+    };
+    /// Program input structure shared pointer type
+    using ProgramInputPtr = std::shared_ptr<ProgramInput>;
+    /// Program input shaded pointer map type
+    using ProgramInputMap = std::unordered_map<std::string, ProgramInputPtr>;
+
+    /// Constructor
+    GlslValidator();
+
+    /// Destructor
+    virtual ~GlslValidator();
+
+    /// @name Setup
+    /// @{
+
+    /// Internal initialization of stages and OpenGL contstructs
+    /// required for program validation and rendering.
+    /// An exception is thrown on failure.
+    /// The exception will contain a list of initialization errors.
+    void initialize();
+
+    /// Return whether the validator has been initialized properly.
+    /// It is possible due to platform ahd machine restrictions that this
+    /// may fail.
+    bool iniitialized() const
+    {
+        return _initialized;
+    }
+
+    /// Set up code stages to validate based on an input hardware shader.
+    /// @param shader Hardware shader to use
+    void setStages(const HwShaderPtr shader);
+
+    /// This method can be used to incrementally set the code stages before validation.
+    /// @param code Shader code string for a given stage
+    /// @param stage Shader stage. See stages allowed in the class HwShader.
+    void setStage(const std::string& code, size_t stage);
+
+    /// Get code string for a given stage
+    /// @return Shader stage string. String is empty if not found.
+    const std::string getStage(size_t stage) const;
+
+    /// Get the number of stages
+    /// @return Stage count
+    size_t numberofStages() const
+    {
+        return HwShader::NUM_STAGES;
+    }
+
+    /// Clear out any existing stages
+    void clearStages();
+
+    /// @}
+    /// @name Program handling
+    /// @{
+
+    /// Create the shader program from stages specified
+    /// An exception is thrown if the program cannot be created.
+    /// The exception will contain a list of program creation errors.
+    /// @return Program identifier. 
+    unsigned int createProgram();
+
+    /// Get list of program input uniforms. 
+    /// The program must have been created successfully first.
+    /// An exception is thrown if the parsing of the program for uniforms cannot be performed.
+    /// @return Program uniforms list.
+    const ProgramInputMap& getUniformsList();
```

Comment:
Can be used outside of validation for apps to perform their own binding vs the "default" binding done by the validator.